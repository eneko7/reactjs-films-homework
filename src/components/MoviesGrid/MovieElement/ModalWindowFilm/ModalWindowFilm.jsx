import React from 'react';
import PropTypes from 'prop-types';
import style from './ModalWindowFilm.scss';

class ModalWindowFilm extends React.Component {
  componentDidMount() {
    const { filmId, fetchFilm } = this.props;
    fetchFilm(filmId);
  }

  render() {
    const { filmId, onChange, filmTrailerKey } = this.props;
    if (filmId > 0) {
      return (
        <div className={style.ModalWindowFilm}>
          <button type="button" className={style.ModalWindowFilm_close} onClick={onChange} />
          <button type="button" className={style.ModalWindowFilm_close_button} onClick={onChange}>&times;</button>
          <iframe
            title="title"
            className={style.ModalWindowFilm_film_iframe}
            src={`https://www.youtube.com/embed/${filmTrailerKey}?autoplay=1&modestbranding=1&showinfo=0&rel=0`}
            frameBorder="0"
            allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
            allowFullScreen
          />
        </div>
      );
    }
    return (
      <div className={style.ModalWindowFilm_film_error} />
    );
  }
}

ModalWindowFilm.propTypes = {
  onChange: PropTypes.func.isRequired,
  fetchFilm: PropTypes.func.isRequired,
  filmId: PropTypes.number.isRequired,
  filmTrailerKey: PropTypes.string.isRequired,
};

export default ModalWindowFilm;
