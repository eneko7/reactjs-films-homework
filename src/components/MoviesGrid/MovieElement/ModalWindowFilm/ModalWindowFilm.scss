.ModalWindowFilm {
  width: 100%;
  top: 0;
  left: 0;
  height: 100%;
  position: fixed;
  z-index: 8;
  display: flex;
  justify-content: center;
  align-items: center;
  background: rgba(0, 0, 0, .5);

  .ModalWindowFilm_close {
    font-size: 70px;
    width: 100%;
    height: 100%;
    z-index: 9;
    color: white;
    position: absolute;
    border: none;
    background: none;
    cursor: pointer;
  }

  .ModalWindowFilm_close_button{
    font-size: 70px;
    z-index: 9;
    color: white;
    position: absolute;
    border: none;
    background: none;
    cursor: pointer;
    right: 50px;
    top: 40px;
  }

  .ModalWindowFilm_close:focus,
  .ModalWindowFilm_close_button:focus {
    outline: none;
  }

  .ModalWindowFilm_film_iframe {
    width: 80%;
    height: 80%;
    position: absolute;
    z-index: 10;
  }
  .ModalWindowFilm_film_error{
    width: 50%;
    height: 50%;
    position: absolute;
    z-index: 10;
    background-image: url(../../../../images/error_video.jpg);
    background-position: center;
    background-size: cover;
    background-repeat: no-repeat;
  }
}