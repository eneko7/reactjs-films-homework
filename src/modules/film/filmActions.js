import axios from 'axios';

export const FETCH_FILM_REQUEST = 'FETCH_FILM_REQUEST';
export const FETCH_FILM_SUCCESS = 'FETCH_FILM_SUCCESS';
export const FETCH_FILM_ERROR = 'FETCH_FILM_ERROR';

export const receiveFilmSuccess = payload => ({
  type: FETCH_FILM_SUCCESS,
  payload,
});

export const receiveFilmRequest = () => ({
  type: FETCH_FILM_REQUEST,
});

export const receiveFilmError = payload => ({
  type: FETCH_FILM_ERROR,
  payload,
});

export function fetchFilm(filmId) {
  return (dispatch) => {
    dispatch(receiveFilmRequest());
    return axios.get(`https://api.themoviedb.org/3/movie/${filmId}/videos?api_key=6a08c0def237c5910708279c9ee78cc5`)
      .then((response) => {
        dispatch(receiveFilmSuccess(response.data.results[0].key));
      })
      .catch(error => dispatch(receiveFilmError(error.message)));
  };
}
