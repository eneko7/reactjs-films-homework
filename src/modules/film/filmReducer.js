import {
  FETCH_FILM_REQUEST,
  FETCH_FILM_SUCCESS,
  FETCH_FILM_ERROR,
} from './filmActions';

const initialState = {
  filmTrailer: '',
  errorFilm: false,
  isFetchingFilm: false,
  isFetchedFilm: false,
};

const filmReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_FILM_SUCCESS:
      return {
        ...state,
        filmTrailer: action.payload,
        errorFilm: false,
        isFetchingFilm: false,
        isFetchedFilm: true,
      };
    case FETCH_FILM_REQUEST:
      return {
        ...state,
        isFetchingFilms: true,
        isFetchedFilms: false,
      };
    case FETCH_FILM_ERROR:
      return {
        ...state,
        errorFilm: true,
      };
    default:
      return {
        ...state,
      };
  }
};

export default filmReducer;
