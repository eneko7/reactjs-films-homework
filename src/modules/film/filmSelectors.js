export const getFilmTrailerKey = state => state.film.filmTrailer;
export const getIsFetchingFilm = state => state.film.isFetchingFilm;
export const getIsErrorFilm = state => state.film.errorFilm;
