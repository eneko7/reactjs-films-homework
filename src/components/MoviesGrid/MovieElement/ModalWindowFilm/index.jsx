import ModalWindowFilm from './ModalWindowFilmContainer';

export default ModalWindowFilm;
