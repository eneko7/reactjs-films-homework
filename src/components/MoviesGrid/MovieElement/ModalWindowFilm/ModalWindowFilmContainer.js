import { connect } from 'react-redux';
import ModalWindowFilm from './ModalWindowFilm';
import { fetchFilm } from '../../../../modules/film/filmActions';
import {
  getFilmTrailerKey,
  getIsFetchingFilm,
  getIsErrorFilm,
} from '../../../../modules/film/filmSelectors';


const mapStateToProps = state => ({
  filmTrailerKey: getFilmTrailerKey(state),
  isFetchingFilm: getIsFetchingFilm(state),
  isErrorFilm: getIsErrorFilm(state),
});

const mapDispatchToProps = {
  fetchFilm,
};

export default connect(mapStateToProps, mapDispatchToProps)(ModalWindowFilm);
