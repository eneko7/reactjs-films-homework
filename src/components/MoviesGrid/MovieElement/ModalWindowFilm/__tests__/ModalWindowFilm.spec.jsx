import React from 'react';
// import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import moxios from 'moxios';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import * as actions from '../../../../../modules/film/filmActions';
import getFilmMock from '../../../../../modules/mocks/getFilmMock';
import ModalWindowFilm from '../index';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({
  film: {
    filmTrailerKey: 'key',
  },
});
const shallow = new ShallowRenderer();
const data = {
  filmId: 1,
  onChange: () => ('Hello!'),
};
const dataFailId = {
  filmId: '',
  onChange: () => ('Hello!'),
};
describe('ModalWindowFilm Snapshot', () => {
  it('renders with trailer', () => {
    const component = shallow.render(
      <Provider store={store}>
        <ModalWindowFilm {...data} />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
});
describe('fetchFilm actions', () => {
  it('renders with trailer', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <ModalWindowFilm {...data} />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
  beforeEach(() => {
    moxios.install();
    store.clearActions();
  });
  afterEach(() => {
    moxios.uninstall();
  });

  it('creates FETCH_FILM_SUCCESS after successfuly fetching film', () => {
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: getFilmMock,
      });
    });
    const expectedActions = [
      { type: actions.FETCH_FILM_REQUEST },
      { type: actions.FETCH_FILM_SUCCESS, payload: getFilmMock.results[0].key },
    ];
    return store.dispatch(actions.fetchFilm()).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
describe('fetchFilm actions', () => {
  it('renders with trailer', () => {
    const component = TestRenderer.create(
      <Provider store={store}>
        <ModalWindowFilm {...dataFailId} />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
  beforeEach(() => {
    moxios.install();
    store.clearActions();
  });
  afterEach(() => {
    moxios.uninstall();
  });

  it('creates FETCH_FILM_SUCCESS after faild fetching film', () => {
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: '',
      });
    });

    const expectedActions = [
      { type: actions.FETCH_FILM_REQUEST },
      { type: actions.FETCH_FILM_ERROR, payload: "Cannot read property '0' of undefined" },
    ];

    return store.dispatch(actions.fetchFilm()).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
