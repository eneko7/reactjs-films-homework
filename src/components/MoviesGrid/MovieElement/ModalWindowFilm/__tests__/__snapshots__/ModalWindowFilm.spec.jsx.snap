// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWindowFilm Snapshot renders with trailer 1`] = `
<Connect(ModalWindowFilm)
  filmId={1}
  filmTrailerKey="key"
  onChange={[Function]}
/>
`;

exports[`ModalWindowFilm Snapshot renders with trailer 2`] = `
<Connect(ModalWindowFilm)
  filmId={-1}
  filmTrailerKey="key"
  onChange={[Function]}
/>
`;

exports[`fetchFilm actions renders with trailer 1`] = `
<div
  className="ModalWindowFilm"
>
  <button
    className="ModalWindowFilm_close"
    onClick={[Function]}
    type="button"
  />
  <button
    className="ModalWindowFilm_close_button"
    onClick={[Function]}
    type="button"
  >
    Ã—
  </button>
  <iframe
    allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
    allowFullScreen={true}
    className="ModalWindowFilm_film_iframe"
    frameBorder="0"
    src="https://www.youtube.com/embed/key?autoplay=1&modestbranding=1&showinfo=0&rel=0"
    title="title"
  />
</div>
`;

exports[`fetchFilm actions renders with trailer 2`] = `
<div
  className="ModalWindowFilm_film_error"
/>
`;
